### PR TITLE
List the resources (by declaration) in a vehicle specification

### PR DESCRIPTION
**Pull request by developing-ar**
# Description of changes
Added the list (top-level) option and sub-command for resources (networks, datasets, parameters), and properties in both a human readable and json format.

Usage:
```bash
$ vehicle list <resources|properties> -s <spec-file> <-j>
```

This is so that these elements can be rendered as "inputs" on the GUI.

### CODE REVIEW

File: vehicle/src/Vehicle/CommandLine.hs [hunk 1]
```diff
@@ -454,6 +455,13 @@ outputAsJSONParser =
       <> help "Output the program as JSON instead of text."
       <> internal
 
+resourcesParser :: Parser Bool
+resourcesParser =
+  switch $
+    long "resources"
+      <> short 'r'
+      <> help "Output the resources outlined in a vehicle file. Only prints if typeCheck is successful."
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/CommandLine.hs]: Maybe "After type-checking output the resources (e.g. networks, parameters, datasets) in the file as JSON"?

File: vehicle/src/Vehicle/TypeCheck.hs [hunk 2]
```diff
@@ -140,6 +143,22 @@ printPropertyTypes (Main decls) = do
       let propertyType = prettyFriendlyEmptyCtx (typeOf decl)
       propertyName <+> ":" <+> propertyType
 
+printResources :: (MonadIO m, MonadCompile m, PrintableBuiltin builtin) => Prog builtin -> m ()
+printResources (Main decls) = do
+  let networks = filterByAbstractSort NetworkDef decls
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/TypeCheck.hs]: So this function definitely works, nice job! I have a couple of suggested improvements though.

1. We'd probably like the resources to be returned in the same order that they were in the file so that they can be displayed in that order in the GUI? 
2. For the GUI, we probably need the type information as well, not just the name, in order to create the right input boxes and do validation?
3. It would be good if we actually output proper JSON rather than hand-crafting a format?

Doing 2 is hard so maybe we'll leave that for another issue, but we should be able to do 1 and 3 pretty easily. Instead of filtering, can you write a function that recurses over all the declarations and returns something of type `[AbstractSort, Name]`. You then should be able to convert that list directly to JSON with `toJSON`?

Also FYI you need to capture `NonInferable` parameters as well.

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Great job! I have a couple of suggestions for improvements, but it would definitely work!

**Comment by MatthewDaggitt** [comment]
Also looks like the help text has changed (see failing test in CI), you will need to update the golden test file. Instructions here: https://github.com/vehicle-lang/vehicle/blob/dev/CONTRIBUTING.md#updating-the-golden-files

**Comment by wenkokke** [comment]
If the semantics of this flag is that it prints the resources and then exits, please transform it into a command named "list-resources".

File: vehicle/src/Vehicle/TypeCheck.hs (hunk 2 above)
Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: @MatthewDaggitt did you intend for that to be a list of tuples?

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: I agree, this looks great! I have a few more suggestions:

The filtering, as well as any other nontrivial operation on the data, should be done externally, rather than in Vehicle, otherwise we'll have to keep modifying Vehicle as your needs change, or we'll end up with the logic split between two places.

The output should be JSON, as it's intended to be consumed programmatically. However, I'd be even happier if the `list-resources` command took an optional `--json` flag that changes its output form human readable to JSON in line with how the other commands work.

Contrary to Matthew's suggestion, I believe the names and sorts of the resources should be output as  a list of objects, so that future additions won't break external tools. This requires a little bit more work, defining a Haskell type and its ToJSON instance.

File: vehicle/src/Vehicle/CommandLine.hs [hunk 3]
```diff
@@ -184,6 +184,7 @@ typeCheckParser =
   TypeCheckOptions
     <$> specificationParser
     <*> typeSystemParser
+    <*> resourcesParser
```
Comment (wenkokke) [inline vehicle/src/Vehicle/CommandLine.hs]: While it requires type checking, `list-resources` should be its own top level command which outputs the resources as well as other info. It should take a `--json` flag that changes its output from a human readable list to JSON.

**Review by wenkokke — CHANGES_REQUESTED** [review]

File: vehicle/src/Vehicle/TypeCheck.hs (hunk 2 above)
Comment (developing-ar) [reply vehicle/src/Vehicle/TypeCheck.hs]: @wenkokke that makes sense for haskell type to have the its ToJSON instance. So, for the 
> names and sorts of the resources should be output as a list of objects

that means that the returned JSON is a "list" of all the resources in the file and each object has its name and "sort" (and later on type).

@MatthewDaggitt type information does not seem to difficult to extract as it is already done in `printProperties`, unless you were expecting a more sophisticated "typing".

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: I think @MatthewDaggitt was thinking about outputting the types as structured JSON. However, it's probably fine to add the printed type as a string, if you'd like to show this to the user. However, it may be a bit tricky to get the type in the right shape and get the right printing function.

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/TypeCheck.hs]: Yes, the type is tricky as we can't convert raw vehicle types of JSON (and nor is it clear we want to). For now, yes we could just `prettyFriendly` the type and attach it to the tuple. @wenkokke is of course right that I meant `[(AbstractSort, Name)]` or now `[(AbstractSort, Name, Text)]` for the type.

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/TypeCheck.hs]: I agree with @wenkokke as well, maybe splitting this out under a separate `resources` command mode is the way to go. (I prefer `resources` over `list-resources`) and having the `--json` flag as well.

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: I think `list-resources` might be better since I anticipate that it will expand to include more list commands, rather than expand to include more resource related commands.

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: > Yes, the type is tricky as we can't convert raw vehicle types of JSON (and nor is it clear we want to). For now, yes we could just `prettyFriendly` the type and attach it to the tuple. @wenkokke is of course right that I meant `[(AbstractSort, Name)]` or now `[(AbstractSort, Name, Text)]` for the type.

It should be a list of a custom type, rather than a tuple, so that we can have a custom ToJSON instance, as mentioned above.

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/TypeCheck.hs]: Ooh like what?

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/TypeCheck.hs]: But okay, yup `list-resources` is fine!

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: > Ooh like what?

```json
{
  "name": "...",
  "sort": "...",
  "type": "..."
}
```

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/TypeCheck.hs]: Sorry, I meant "What other things will we want to list in the future?"

Comment (wenkokke) [reply vehicle/src/Vehicle/TypeCheck.hs]: Properties. Functions. Implicit variables. Who knows?

File: vehicle/src/Vehicle/CommandLine.hs (hunk 3 above)
Comment (developing-ar) [reply vehicle/src/Vehicle/CommandLine.hs:187]: Sounds good. Will do!

**Comment by developing-ar** [comment]
@wenkokke @MatthewDaggitt then, how about a new top-level command `list` and then `--resources`, `--properties` etc. (and `--json`) can all be command-line options?

**Comment by MatthewDaggitt** [comment]
> @wenkokke @MatthewDaggitt then, how about a new top-level command `list` and then `--resources`, `--properties` etc. (and `--json`) can all be command-line options?

Yup, sounds good to me!

**Comment by wenkokke** [comment]
> @wenkokke @MatthewDaggitt then, how about a new top-level command `list` and then `--resources`, `--properties` etc. (and `--json`) can all be command-line options?

I'd prefer them to be subcommands, in that case, because they should be exclusive, so:

`vehicle list resources`

and:

`vehicle list resources --json`

**Comment by developing-ar** [comment]
> > @wenkokke @MatthewDaggitt then, how about a new top-level command `list` and then `--resources`, `--properties` etc. (and `--json`) can all be command-line options?
> 
> I'd prefer them to be subcommands, in that case, because they should be exclusive, so:
> 
> `vehicle list resources`
> 
> and:
> 
> `vehicle list resources --json`

Done! Although its `Resources` instead of `resources`. I can update that if the case is important. I've added a `Properties` option as well.

**Comment by wenkokke** [comment]
The case is important. All commands should be lowercase, for various reasons. Please do change it to lowercase.

**Comment by developing-ar** [comment]
> The case is important. All commands should be lowercase, for various reasons. Please do change it to lowercase.

@wenkokke Sorry, I forgot to clarify that I'm using it as an option rather than a sub-command.
For example, in typeCheck:
```
-t <Standard|Linearity|Polarity>
```
in List:
```
-l <Resources|Properties>
```
I can convert it to a sub-command instead, and use the lower case accordingly - unless it's fine to have them as options.

**Comment by wenkokke** [comment]
> > The case is important. All commands should be lowercase, for various reasons. Please do change it to lowercase.
> 
> 
> 
> @wenkokke Sorry, I forgot to clarify that I'm using it as an option rather than a sub-command.
> 
> For example, in typeCheck:
> 
> ```
> 
> -t <Standard|Linearity|Polarity>
> 
> ```
> 
> in List:
> 
> ```
> 
> -l <Resources|Properties>
> 
> ```
> 
> I can convert it to a sub-command instead, and use the lower case accordingly - unless it's fine to have them as options.

I'd prefer as lowercase commands. Firstly, because always lowercase. Secondly, because I'd ideally like options to have globally unique names (ie for every subcommand) to avoid confusion, and would prefer not to push us further from that ideal.

File: vehicle/src/Vehicle/List.hs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  { entitySort :: Text,
+    entityName :: Text,
+    entityType :: Text
+  }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON ListEntity
+
+instance Pretty ListEntity where
+  pretty listEntity = pretty (entitySort listEntity) <+> pretty (entityName listEntity) <+> pretty (entityType listEntity)
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/List.hs]: Can we avoid over long lines, maybe:
```haskell
pretty listEntity = 
  pretty (entitySort listEntity) <+> 
  pretty (entityName listEntity) <+> 
  pretty (entityType listEntity)
```

File: vehicle/src/Vehicle/List.hs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  (imports, typedProg) <-
+    typeCheckUserProg $
+      TypeCheckOptions
+        { specification = specification,
+          typingSystem = Standard
+        }
+  let mergedProg = mergeImports imports typedProg
+  printResources mergedProg listEntities outputAsJSON
+
+printResources :: (MonadIO m, MonadCompile m, PrintableBuiltin builtin) => Prog builtin -> ListableEntities -> Bool -> m ()
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/List.hs]: Again can we lay this out as:
```haskell
printResources :: 
  (MonadIO m, MonadCompile m, PrintableBuiltin builtin) => 
  Prog builtin -> 
  ListableEntities -> 
  Bool -> 
  m ()
```

File: vehicle/src/Vehicle/List.hs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+printResources (Main decls) listEntities outputAsJSON = do
+  let filteredDecls =
+        if listEntities == ExternalResources
+          then filter isAbstractDecl decls
+          else filter isPropertyDecl decls
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/List.hs]: This could be written more succinctly as:
```haskell
let filterFn = case listEntities of
  ExternalResources -> isAbstractDecl
  Properties -> isPropertyDecl
let filteredDecls = filter filterFn decls
```
This has two advantages. Firstly reduces the duplication of `filter ... decls`. Secondly the compiler will warn if we add a new option to `ListableEntities`.

File: vehicle/src/Vehicle/List.hs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  let listDecls = fmap (\decl -> convertDeclToListEntity decl (if listEntities == ExternalResources then pack $ show $ pretty (abstractSortOf decl) else "@property")) filteredDecls
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/List.hs]: I don't think you need the `pack (show)` here as you are pretty printing it on line 42.

File: vehicle/src/Vehicle/List.hs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  let outputDocs =
+        if outputAsJSON
+          then pretty $ unpack $ encodePretty' prettyJSONConfig $ toJSON listDecls
+          else pretty listDecls
```
Comment (MatthewDaggitt) [inline vehicle/src/Vehicle/List.hs:57]: Can extract the `pretty` in both branches to line 46.

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Looks good! Some minor Haskell style pointers!

**Comment by MatthewDaggitt** [comment]
Yup, a subcommand would be great!

File: vehicle/src/Vehicle/List.hs (hunk 6 above)
Comment (developing-ar) [reply vehicle/src/Vehicle/List.hs]: Great! will do

File: vehicle/src/Vehicle/List.hs (hunk 7 above)
Comment (developing-ar) [reply vehicle/src/Vehicle/List.hs]: It needs to be in `Text` for `convertDeclToListEntity()` but I've removed that and made the `convertDeclToListEntity` use `Doc a` instead.

File: vehicle/src/Vehicle/List.hs (hunk 8 above)
Comment (developing-ar) [reply vehicle/src/Vehicle/List.hs:57]: `pretty` is applied differently for each branch (produced  types), and extracting `pretty` gives the error for `listDecls`
```
 Couldn't match type ‘ListEntity’ with ‘Char’
```
Unless I'm missing something...

Comment (MatthewDaggitt) [reply vehicle/src/Vehicle/List.hs:57]: Ah right, fair enough!

**Review by MatthewDaggitt — COMMENTED** [review]
Looks good to me! Can you add a couple of golden tests for example specifications that have resources (e.g. mnist-robustness, acasXu)

**Comment by developing-ar** [comment]
@MatthewDaggitt  I'm adding golden tests for acasXu, mnist-robustness, and wind-controller for both `list resources` and `list properties`. Would you like me to update all the tests for **all** the golden tests (examples) ?

File: vehicle/tests/golden/compile/acasXu/ListProperties.out.golden [hunk 9]
```diff
@@ -0,0 +1,10 @@
+[ @property property1 Bool
```
Comment (MatthewDaggitt) [inline vehicle/tests/golden/compile/acasXu/ListProperties.out.golden]: Hmm, I was kind of expecting that this was going to look more like JSON:
```
[ { entitySort = @property,
 , entityName = property1,
 , entityType = Bool
} ,
...
]
```
This current format is going to be kind of hard to parse on the other side!

**Review by MatthewDaggitt — COMMENTED** [review]
Yes this small set of tests is fine. It's not terribly complicated code at the end of the day, and I think you've covered all the edge cases.

It would be good if we could tweak the actual format to be JSON, but after that happy to merge in!

File: vehicle/tests/golden/compile/acasXu/ListProperties.out.golden (hunk 9 above)
Comment (developing-ar) [reply vehicle/tests/golden/compile/acasXu/ListProperties.out.golden]: The JSON option does give this output format which should be fine to parse:
```
[
  {
    "entitySort": "@property",
    "entityName": "property1",
    "entityType": "Bool",
  },
 ...
]
```
I did not include the json option in the tests, I'll update that - I assumed the json option might not be part of the golden tests since it's not the "base" output, since a previous [discussion](https://github.com/vehicle-lang/vehicle/pull/885#discussion_r2034454551) was that the default output is a human-readable list, and the optional json option.

Unless you meant for changing the default format?

Comment (MatthewDaggitt) [reply vehicle/tests/golden/compile/acasXu/ListProperties.out.golden]: No, that's perfect! As this API is meant to be consumed by another program rather than a human, I think it makes most sense to have the golden test use JSON!

Thanks for the change!

**Review by MatthewDaggitt — APPROVED** [review]
I'm happy with it! @wenkokke ?

**Review by wenkokke — APPROVED** [review]

**Comment by wenkokke** [comment]
Looks good to me!
